### ISSUE TITLE
Stylistic conventions on function names and underscores

### ISSUE THREAD

**Issue by wbhart** [JuliaLang/julia#15547]
Under variables in the Julia manual [1], it suggests that "word" separation be done using underscores if it makes it more readable but later says flat out that function names should not contain underscores.

However, in the actual style guide [2] it says that function names should have underscores if it makes them more readable.

The contributors guide suggests functions should always have underscores [3].

This ticket is not about what the convention should be, as I think it is clear if you read all three locations, but merely to point out the apparent contradiction on a first reading of each individual statement.

BTW, none of the three says anything about the convention on leading underscores on function and variable names.

[1] http://docs.julialang.org/en/latest/manual/variables/#stylistic-conventions

[2] http://docs.julialang.org/en/latest/manual/style-guide/#use-naming-conventions-consistent-with-julia-s-base

[3] https://github.com/JuliaLang/julia/blob/31b04bde2d01c8ee23fb4160239fc30afb4b4792/CONTRIBUTING.md#general-formatting-guidelines-for-julia-code-contributions

**Comment by johnmyleswhite** [comment]
I don't think this issue is likely to go anywhere, but I agree that this issue needs resolving at some point. Our current de facto rules are pretty subjective and, IMO, lack a ["bright line"](https://en.wikipedia.org/wiki/Bright-line_rule).

**Comment by wbhart** [comment]
That's essentially why I suggested the ticket is not about deciding a
convention, but merely eliminating the ambiguity in the documentation.

I think the currently deliberately vague, de facto convention seems to be
"function names are lower case, without underscores, except when separating
words with underscores will make them more readable, e.g. isequal and
sha_request".

I agree that the problem of defining a "bright line" is a much harder
problem in Computer Science and could easily result in bikeshedding. I
certainly wasn't looking to start such a discussion here.
